Fix: 3DS modal does not appear for test cards for recurring payment in India in non-Indian account (2)

Mayisha opened the pull request:
Fixes STRIPE-341 #4062 

When the store is connected to a non-indian Stripe account and has a different currency than INR, the 3DS modal is not working when paid using [test cards for recurring payment in India](https://href.li/?https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents#testing) for a free trial subscription. This happens for a subscription with free trials only, basically when the total amount is 0. The 3DS modal appears fine if the subscription needs to be charged (has a non zero total price).

First PR: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4094
Revert PR: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4190
We had to revert it because the changes caused regression and the regular checkout flow was broken. Apparently, during setup intent creation, we need to set `currency` in `mandate_option` but for payment intent this is an unsupported property. 

## Changes proposed in this Pull Request:
- Added the `wc_stripe_generate_create_intent_request` filter in [create_and_confirm_setup_intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-intent-controller.php#L1030) function to make sure mandate information is added to the request. If we do not include the mandate data to the setup intent request, the 3DS modal does not appear.
- Added currency data to the `mandate_option`. This is required for setup intent but unsupported for payment intent. 

## Testing instructions

- Connect to a non-Indian Stripe account.
- Set the store currency to something except INR.
- Create a subscription product with a free trial.
- As a shopper, add the free trial subscription to your cart and go to the checkout page.
- In `develop` try to pay with the card `4000003560000123`.
- Notice that when you click checkout, it directly takes you to the order received page without showing the 3DS authentication modal.
- Checkout to this branch and try to pay with the card `4000003560000123` for the same free subscription product.
- Confirm that the 3DS modal is presented.
- Regression test
    - Test and confirm that purchasing a non-subscription product with a normal card works.
    - Test and confirm that purchasing a non-subscription product with a 3DS card works.
    - Test and confirm that purchasing a subscription product (non-zero price) with a normal card works.
    - Test and confirm that purchasing a subscription product (non-zero price) with a 3DS card works.    
    - Confirm that you can save payment methods from `My Account` page.
    - Connect to an Indian Stripe account, set your store currency to INR. Purchase a subscription using [test cards for recurring payment in India](https://href.li/?https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents#testing). Check your logs and find the `process response` log for this order. Confirm that there is a mandate id present in the response object.

Mayisha committed:
apply 'wc_stripe_generate_create_intent_request' flter when creating a setup intent for card

Mayisha committed:
add tests

Mayisha committed:
add changelog

Mayisha committed:
fix lint

Mayisha committed:
Merge branch 'develop' into fix/4062-3ds-indian-recurring-non-indian-account-2

annemirasol reviewed (changes requested):
It looks like Pre-orders is broken in this branch:

<img width="741" alt="Screenshot 2025-04-08 at 4 06 54 PM" src="https://github.com/user-attachments/assets/805a21b1-4f36-4c29-b98e-5c6593bc1453" />

Mayisha committed:
Merge branch 'develop' into fix/4062-3ds-indian-recurring-non-indian-account-2

Mayisha committed:
add a helper method to check if currency is supported for indian recurring payment mandate

Mayisha committed:
fix for pre orders

Mayisha committed:
refactor in subscription trait
- use the helper method for supported currency check
- merge existing mandate option data with the new values

Mayisha committed:
add fallback for order key

Mayisha committed:
fix lint

Mayisha committed:
Merge branch 'develop' into fix/4062-3ds-indian-recurring-non-indian-account-2

Mayisha committed:
Merge branch 'develop' into fix/4062-3ds-indian-recurring-non-indian-account-2

annemirasol commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -825,10 +825,11 @@ public function create_and_confirm_payment_intent( $payment_information ) {
 	 * @param array       $request              The request array to add the mandate options to.
 	 * @param string|null $payment_method_type  The type of payment method to use for the intent.
 	 * @param bool        $is_setup_intent      Whether the request is for a setup intent.
+	 * @param WC_Order    $order                The order object.
```
Nit: `WC_Order|null`

annemirasol reviewed (approved):
✅ Test and confirm that purchasing a non-subscription product with a normal card works.
✅ Test and confirm that purchasing a non-subscription product with a 3DS card works.
✅ Test and confirm that purchasing a subscription product (non-zero price) with a normal card works.
✅ Test and confirm that purchasing a subscription product (non-zero price) with a 3DS card works.
✅ Confirm that you can save payment methods from My Account page.
✅ Connect to an Indian Stripe account, set your store currency to INR. Purchase a subscription using [test cards for recurring payment in India](https://href.li/?https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents#testing). Check your logs and find the process response log for this order. Confirm that there is a mandate id present in the response object.
✅ Pre-orders work

Documenting here, but also seeing these in `develop` when I checked. I will create issues for them:
- Getting a PHP notice when purchasing a free trial sub with a saved payment method (normal card). (EDIT: STRIPE-404)
 
```
PHP Notice:  Trying to get property 'type' of non-object in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php on line 1998
```


- Getting a PHP notice when saving a payment method: (EDIT: goes away after a Stripe settings save)
```
[14-Apr-2025 02:39:37 UTC] PHP Notice:  Undefined index: single_payment_element in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php on line 371
```

Mayisha committed:
fix param type
